Add MultiLineLogFormatter

BCSol opened the pull request:
MultiLineLogFormatter that supports color mode when attached to console.
Prints timestamp, log level, first log message line, and component
description in this order as first line. Appends remaining log message
lines as additional lines.

Make the LogFormatter used in the VerifierCloud available for the witness-browser.

Commit:
Add TimestampedLogFormatter with color support

Refactored colouring functionality of ConsoleLogFormatter into abstract
class.
TimestampedLogFormatter replaces LogFileFormatter with color support.
Marked FileLogFormatter as deprecated.

PhilippWendler commented on src/org/sosy_lab/common/log/MultiLineLogFormatter.java:
```diff
@@ -0,0 +1,75 @@
+package org.sosy_lab.common.log;
```
License header is missing.

BCSol replied:
- added licence header
- added static factory methods, but no `LoggingOption`s constructor
- Refactored coloured format feature into abstract class

We should maybe come up with a better name for the formatter. I named it `MultilineLogFormatter`, because the first line of the logmessage is displayed inline with:

`timestamp` `level` `message` [`component`:`class`.`method`]   
`additional optional message lines`

PhilippWendler replied:
Hm, that does not really differ from the behavior of the other formatter, right? It also shows all lines of the message.

Is the main difference thus the timestamp? Then the name could be based on this.
Can other differences be eliminated?

BCSol replied:
The main difference is in fact the timestamp, so this could be an option.

The other difference is that we wanted to have the most important information in chronological order on the first line:
- timestamp
- level: if the log message becomes very long, you have to look for the importance  
- message: only the first line, usually contains message, more lines can be seen as additional information, such as the stack trace
- component: since the vcloud doesn't fully leverage that feature yet and due to it's length it was moved to the end.

So the difference to the `FileLogFormatter` is only the position of the component descriptor, but it differs greatly to the `ConsoleLogFormatter`. We use this formatter for both the file and console logging. 
Technically we could refactor `FileLogFormatter` to allow the usage of colours as well and move to it. But then we should rename that one to TimestampedLogFormatter?

BCSol replied:
We could generalize it completely, as well, so that users can provide a formatting string like:
`{timestamp:yyyy-MM-dd HH:mm:ss:SSS} {level} {message0} ({component}:{method})\n{message1+}`

PhilippWendler replied:
> Technically we could refactor `FileLogFormatter` to allow the usage of colours as well and move to it. But then we should rename that one to TimestampedLogFormatter?

That would be nice. The current `FileLogFormatter` would then stay, but be changed to extend `TimestampedLogFormatter` and marked as `@Deprecated`.

But I note that the current code of `TimestampedLogFormatter` is still different from `FileLogFormatter`. Do you intend to change that?

I would prefer to not invent our own customization mechanism for the message. We do not want to implement our own logging framework.

BCSol replied:
>But I note that the current code of TimestampedLogFormatter is still different from FileLogFormatter. Do you intend to change that?

I guess the vcloud can live with the changed order of log entries and there might even be a incentive to you use the component logging mechanism more.

>That would be nice. The current FileLogFormatter would then stay, but be changed to extend TimestampedLogFormatter and marked as @Deprecated.

Sounds good, I'll make the necessary changes then.

PhilippWendler reviewed (changes requested):
This duplicates too much code from `ConsoleLogFormatter`. Please refactor.

Furthermore, please add static factory methods.

PhilippWendler commented on src/org/sosy_lab/common/log/AbstractColoredLogFormatter.java:
```diff
@@ -0,0 +1,79 @@
+ * Abstract class for creating {@link Formatter}s that color {@link LogRecord}s with {@link
+ * Level#SEVERE} and {@link Level#WARNING} red.
+ */
+public abstract class AbstractColoredLogFormatter extends Formatter {
```
Please remove `public`.

PhilippWendler commented on src/org/sosy_lab/common/log/AbstractColoredLogFormatter.java:
```diff
@@ -0,0 +1,79 @@
+  }
+
+  @Override
+  public String format(LogRecord lr) {
```
Should be `final` because subclasses are not supposed to override it, right?

PhilippWendler commented on src/org/sosy_lab/common/log/AbstractColoredLogFormatter.java:
```diff
@@ -0,0 +1,79 @@
+    return MoreObjects.toStringHelper(this).add("useColors", useColors).toString();
+  }
+
+  protected abstract void format(LogRecord lr, StringBuilder sb);
```
This should have JavaDoc explaining what subclasses are supposed to do.

PhilippWendler reviewed (changes requested):
Please update with `master` (or rebase) and adjust the license header with the current template. Sorry for the inconvenience.

PhilippWendler commented on src/org/sosy_lab/common/log/ConsoleLogFormatter.java:
```diff
@@ -8,61 +8,32 @@
-      }
-    }
-    useColors = pUseColors;
+  private ConsoleLogFormatter(@Var boolean useColors) {
```
Use `@Var` only if the variable is going to me modified (it is the inverse of `final`).

PhilippWendler commented on src/org/sosy_lab/common/log/TimestampedLogFormatter.java:
```diff
@@ -0,0 +1,68 @@
+          .withLocale(Locale.getDefault(Locale.Category.FORMAT))
+          .withZone(ZoneId.systemDefault());
+
+  private TimestampedLogFormatter(@Var boolean useColors) {
```
Same here.

PhilippWendler committed:
Add missing final for field

PhilippWendler committed:
Rewrite JavaDoc to conform to standard.

PhilippWendler committed:
Remove erroneous overload of toString() in log formatter.

The super class already defines toString(),
and by overwriting it we hide the useColors value in the output.

PhilippWendler reviewed (approved):
I added a few small improvements. In the future, please push to a branch in this repository for creating a PR, this will give us a working CI for the PR. (Furthermore, creating a PR from master is discouraged.)
